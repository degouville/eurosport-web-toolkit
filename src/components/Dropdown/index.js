import React from 'react';
import PropTypes from 'prop-types';
import styled, { css } from 'react-emotion';
import Chevron from 'src/assets/chevron.component.svg';
import { medium } from '../../breakpoints';
import { H3 } from '../../typography';

export const StyledContainer = styled.div`
  display: flex;
  align-items: center;
  cursor: pointer;
`;

export const StyledOverlay = styled.div`
  position: fixed;
  width: 100vw;
  height: 100vh;
  top: 0;
  left: 0;
  z-index: 1;

  display: none;
  ${({ isDropdownOpen }) =>
    isDropdownOpen &&
    css`
      display: block;
    `}
`;

export const StyledWrapper = styled.div`
  ${medium(css`
    width: fit-content;
  `)};
  position: relative;
`;

export const StyledText = styled(H3)`
  color: ${({ theme }) => theme.dropdown.label.color};
  text-transform: uppercase;
`;

export const StyledArrow = styled(Chevron)`
  height: 12px;
  width: 12px;
  transform: rotate(90deg);
  margin-left: 5px;
  path {
    fill: ${({ theme }) => theme.dropdown.label.color};
  }
`;

export const StyledDropdown = styled.div`
  position: absolute;
  width: calc(100% - 25px);
  top: 45px;
  ${medium(css`
    right: -94px;
    @-moz-document url-prefix() {
      right: -91px;
    }
    @supports (-ms-ime-align: auto) {
      right: unset;
    }
    bottom: auto;
    width: fit-content;
    max-width: 183px;
  `)};
  display: none;
  z-index: 2;
  opacity: 0.95;
  align-self: flex-end;
  border: solid 1px ${({ theme }) => theme.dropdown.border};
  border-radius: 3px;
  background-color: ${({ theme }) => theme.dropdown.background};
  padding: 17px 8px 12px 17px;
  ${({ isDropdownOpen }) =>
    isDropdownOpen &&
    css`
      display: block;
    `}
  &::before {
    content: '';
    position: absolute;
    width: 10px;
    border-top: 1px solid ${({ theme }) => theme.dropdown.border};
    border-left: 1px solid ${({ theme }) => theme.dropdown.border};
    border-radius: 1px;
    background: ${({ theme }) => theme.dropdown.background};
    height: 10px;
    display: block;
    margin-left: 50%;
    transform: translateX(-50%) rotate(45deg);
    top: -6px;
    left: 0;
  }
`;

const StyledUl = styled.ul`
  height: 130px;
  overflow-y: scroll;
  ::-webkit-scrollbar {
    width: 6px;
  }

  ::-webkit-scrollbar-track {
    background: ${({ theme }) => theme.scrollbar.background};
    border-radius: 2px;
  }

  ::-webkit-scrollbar-thumb {
    background: ${({ theme }) => theme.scrollbar.color};
    border-radius: 2px;
  }
`;

const StyledLi = styled.li`
  color: ${({ isSelected, theme }) =>
    isSelected ? theme.dropdown.list.selected.color : theme.dropdown.list.default.color};
  font-size: 14px;
  cursor: pointer;
  padding: 13px 0;
  margin-right: 10px;
  display: flex;
  justify-content: space-between;
  &:first-of-type {
    padding-top: 0;
  }
  &:not(:last-of-type) {
    border-bottom: solid 1px ${({ theme }) => theme.dropdown.border};
  }
  &:hover {
    color: ${({ theme }) => theme.dropdown.list.selected.color};
  }
`;

export const StyledCheckMark = styled.span`
  color: ${({ theme }) => theme.dropdown.checkMark.color};
  margin: 0 0 0 1em;
`;

const StyledP = styled.p`
  width: 100%;
  ${medium(css`
    width: 130px;
  `)};
`;

class Dropdown extends React.Component {
  state = {
    isDropdownOpen: false,
    selectedOption: null,
  };

  componentWillMount() {
    this.setSelectedOption();
  }

  setSelectedOption = () => {
    const { options } = this.props;
    this.setState({ selectedOption: options[0] });
  };

  onOptionClick = id => {
    const { options, onItemSelected } = this.props;
    const matchingOption = options.find(option => option.id === id);
    onItemSelected && onItemSelected(matchingOption);

    if (matchingOption) {
      this.setState({ selectedOption: matchingOption, isDropdownOpen: false });
    }
  };

  toggleOpen = () => {
    // eslint-disable-next-line no-unused-vars
    this.setState(state => ({
      isDropdownOpen: !state.isDropdownOpen,
    }));
  };

  render() {
    const { options } = this.props;
    const { isDropdownOpen, selectedOption } = this.state;
    return (
      <StyledWrapper>
        <StyledOverlay isDropdownOpen={isDropdownOpen} onClick={this.toggleOpen} />
        <StyledDropdown isDropdownOpen={isDropdownOpen}>
          <StyledUl>
            {options &&
              options.map(option => (
                <StyledLi
                  onClick={() => this.onOptionClick(option.id)}
                  isSelected={selectedOption.id === option.id}
                  key={option.id}
                >
                  <StyledP>{option.text}</StyledP>
                  {selectedOption.id === option.id && <StyledCheckMark> âœ“</StyledCheckMark>}
                </StyledLi>
              ))}
          </StyledUl>
        </StyledDropdown>
        <StyledContainer onClick={this.toggleOpen}>
          <StyledText>
            {selectedOption.text}
            <StyledArrow />
          </StyledText>
        </StyledContainer>
      </StyledWrapper>
    );
  }
}

Dropdown.defaultProps = {
  onItemSelected: null,
};

Dropdown.propTypes = {
  options: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.number,
      text: PropTypes.string,
    })
  ).isRequired,
  onItemSelected: PropTypes.func,
};

export default Dropdown;
