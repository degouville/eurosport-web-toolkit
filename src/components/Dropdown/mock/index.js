export const dropdownOptions = [
  { text: 'Men singles', id: 1231 },
  { text: 'Women singles', id: 1232 },
  { text: 'Men doubles', id: 1233 },
  { text: 'Women doubles', id: 1234 },
  { text: 'Men triples', id: 1235 },
  { text: 'Women triples', id: 1236 },
];

export const initialOptionID = dropdownOptions[3].id;

export default {
  ...dropdownOptions,
  ...initialOptionID,
};
