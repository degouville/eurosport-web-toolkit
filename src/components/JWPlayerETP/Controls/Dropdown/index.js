import React from 'react';
import styled from 'react-emotion';
import PropTypes from 'prop-types';

const Dropdown = ({ children, bottomDisplay }) => (
  <MainContainer>
    <StyledContainer bottomDisplay={bottomDisplay}>
      <StyledDropdownPick bottomDisplay={bottomDisplay} />
      <StyledDropdown>{children}</StyledDropdown>
    </StyledContainer>
  </MainContainer>
);

Dropdown.propTypes = {
  children: PropTypes.node.isRequired,
  bottomDisplay: PropTypes.bool.isRequired,
};

const MainContainer = styled.div`
  position: relative;
  display: flex;
  flex: 1;
  justify-content: center;
  align-items: center;
`;

const StyledContainer = styled.div`
  position: absolute;
  display: flex;
  flex: 1;
  justify-content: center;
  align-items: center;
  flex-direction: ${({ bottomDisplay }) => (bottomDisplay ? 'column-reverse' : 'column')};
  ${({ bottomDisplay }) => bottomDisplay && 'bottom: 100%;'}
`;

const StyledDropdownPick = styled.div`
  border: solid 1px ${({ theme }) => theme.playerControls.dropdown.border};
  ${({ bottomDisplay }) => bottomDisplay && 'border-left: none;'}
  ${({ bottomDisplay }) => bottomDisplay && 'border-top: none;'}
  ${({ bottomDisplay }) => !bottomDisplay && 'border-right: none;'}
  ${({ bottomDisplay }) => !bottomDisplay && 'border-bottom: none;'}
   
  border-radius: 3px;
  background-color: ${({ theme }) => theme.playerControls.dropdown.background};
  height: 10px;
  width: 10px;
  transform: ${({ bottomDisplay }) => (bottomDisplay ? ' translate(0%, -50%)' : 'translate(0%, 50%)')} rotate(45deg);
`;

export const StyledDropdown = styled.div`
  border: solid 1px ${({ theme }) => theme.playerControls.dropdown.border};
  border-radius: 3px;
  background-color: ${({ theme }) => theme.playerControls.dropdown.background};
  padding: 18px;
`;

export default Dropdown;
