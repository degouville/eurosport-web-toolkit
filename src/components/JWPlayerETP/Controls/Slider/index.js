import React from 'react';
import RCSlider from 'rc-slider';
import PropTypes from 'prop-types';

const getHandleStyle = ({ handleSize, railThickness, handleColor }) => ({
  borderColor: 'transparent',
  height: handleSize,
  width: handleSize,
  backgroundColor: handleColor,
  margin: -(handleSize / 2) + railThickness / 2,
});

const getBarStyle = ({ backgroundColor, railThickness, vertical }) => {
  if (vertical) return { backgroundColor, width: railThickness };
  return { backgroundColor, height: railThickness };
};

const Slider = ({
  value,
  onChange,
  defaultValue,
  min,
  max,
  vertical,
  railThickness,
  railColor,
  handleSize,
  handleColor,
  trackColor,
}) => (
  <RCSlider
    defaultValue={defaultValue}
    onChange={onChange}
    vertical={vertical}
    trackStyle={getBarStyle({ backgroundColor: trackColor, railThickness, vertical })}
    handleStyle={getHandleStyle({ handleSize, railThickness, handleColor })}
    railStyle={getBarStyle({ backgroundColor: railColor, railThickness, vertical })}
    value={value}
    min={min}
    max={max}
  />
);

Slider.propTypes = {
  // Values
  // eslint-disable-next-line react/forbid-prop-types
  min: PropTypes.number.isRequired,
  max: PropTypes.number.isRequired,
  value: PropTypes.number.isRequired,
  onChange: PropTypes.func.isRequired,
  defaultValue: PropTypes.number.isRequired,

  // Theme
  vertical: PropTypes.bool.isRequired,
  railThickness: PropTypes.number.isRequired,
  railColor: PropTypes.string.isRequired,
  handleSize: PropTypes.number.isRequired,
  handleColor: PropTypes.string.isRequired,
  trackColor: PropTypes.string.isRequired,
};

export default Slider;
