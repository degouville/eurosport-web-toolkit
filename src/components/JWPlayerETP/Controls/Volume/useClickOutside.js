import { useLayoutEffect } from 'react';

const useClickOutside = ({ ref, callback }) =>
  useLayoutEffect(() => {
    const listener = event => ref.current && !ref.current.contains(event.target) && callback();
    window.addEventListener('click', listener);
    return () => window.removeEventListener('click', listener);
  }, [ref, callback]);

export default useClickOutside;
