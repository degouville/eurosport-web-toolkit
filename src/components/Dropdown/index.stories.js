import React from 'react';
import { storiesOf } from '@storybook/react';
import { object, number } from '@storybook/addon-knobs';
import { Dropdown } from 'src';
import { dropdownOptions, initialOptionID } from './mock';

const dropdownStories = storiesOf('Components|Dropdown', module);

dropdownStories
  .add('Default', () => <Dropdown options={object('options', dropdownOptions)} />)
  .add('Default with 1 item', () => <Dropdown options={object('options', dropdownOptions.slice(0, 1))} />)
  .add('With Initial Option', () => (
    <Dropdown
      options={object('options', dropdownOptions)}
      initialOptionID={number('InitialOptionID', initialOptionID)}
    />
  ));
