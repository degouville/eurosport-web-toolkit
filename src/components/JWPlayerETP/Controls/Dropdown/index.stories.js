import React from 'react';
import { storiesOf } from '@storybook/react';
import theme from 'src/theme';
import { boolean } from '@storybook/addon-knobs';
import styled from 'react-emotion';
import Dropdown from './index';
import Slider from '../Slider';

const dropdownStories = storiesOf('Components|PlayerDropdown', module);

dropdownStories.add('Default', () => (
  <MainContainer>
    <Dropdown theme={theme.playerControls.dropdown} bottomDisplay={boolean('bottomDisplay', true)}>
      <SliderContainer>
        <Slider max={100} onChange={() => null} min={0} value={50} defaultValue={0} {...theme.playerControls.volume} />
      </SliderContainer>
    </Dropdown>
  </MainContainer>
));

const MainContainer = styled.div`
  margin-top: 200px;
  display: flex;
  flex: 1;
  flex-direction: column;
  align-items: center;
`;

const SliderContainer = styled.div`
  height: 100px;
`;
