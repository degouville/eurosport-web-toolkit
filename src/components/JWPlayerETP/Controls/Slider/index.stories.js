import React, { useState } from 'react';
import { storiesOf } from '@storybook/react';
import theme from 'src/theme';
import Slider from './index';

const stories = storiesOf('Components|JWPlayerETP', module);

const SliderControl = props => {
  const [progress, setProgress] = useState(50);
  return <Slider value={progress} onChange={setProgress} defaultValue={50} {...props} />;
};

stories.add('Slider horizontal', () => <SliderControl {...theme.playerControls.seek} />);
stories.add('Slider vertical', () => (
  <div style={{ width: 500, height: 500 }}>
    <SliderControl {...theme.playerControls.volume} />
  </div>
));
