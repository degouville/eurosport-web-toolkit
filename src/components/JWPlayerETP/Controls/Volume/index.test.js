import React from 'react';
import { shallow } from 'enzyme';
import VolumeOff from 'src/assets/icon-volume-off.svg';
import VolumeMedium from 'src/assets/icon-volume-medium.svg';
import VolumeHigh from 'src/assets/icon-volume-high.svg';
import Icon from '../UI/icon';
import { BarWithPointer, DropdownContainer, Volume } from './index';
import Slider from '../Slider';
import Dropdown from '../Dropdown';
import useClickOutside from './useClickOutside';

jest.mock('./useClickOutside');

describe('Volume', () => {
  const createDefaultProps = newProps => ({
    volume: 50,
    onVolume: jest.fn(),
    mute: false,
    theme: {
      playerControls: {
        volume: {
          trackColor: 'blue',
          handleSize: 20,
          handleColor: 'red',
          railColor: 'green',
          railThickness: 5,
          vertical: false,
        },
      },
    },
    ...newProps,
  });

  describe('Snapshots', () => {
    it('Should match snapshot when the volume is not shown', () => {
      // Given
      const props = createDefaultProps();
      const wrapper = shallow(<Volume {...props} />);

      // Expect
      expect(wrapper).toMatchSnapshot();
    });
  });

  describe('Volume bar', () => {
    beforeEach(() => {
      useClickOutside.mockImplementation(() => jest.fn());
    });

    afterEach(() => {
      jest.resetAllMocks();
    });

    it('Should contains Dropdown when the volume is shown', () => {
      // Given
      const props = createDefaultProps();
      const wrapper = shallow(<Volume {...props} />);

      // When
      wrapper.find(BarWithPointer).simulate('click');
      const result = wrapper.find(Dropdown).exists();

      // Expect
      expect(result).toBe(true);
    });

    it('Should not contains Dropdown when the volume is shown', () => {
      // Given
      const props = createDefaultProps();
      const wrapper = shallow(<Volume {...props} />);

      // When
      wrapper.find(BarWithPointer).simulate('click');
      useClickOutside.mock.calls[1][0].callback();
      wrapper.update();
      const result = wrapper.find(Dropdown).exists();

      // Expect
      expect(result).toBe(false);
    });
  });

  describe('Volume icon', () => {
    it('Should show volume off when the player is muted', () => {
      // Given
      const props = createDefaultProps({ mute: true });
      const wrapper = shallow(<Volume {...props} />);

      // When
      const source = wrapper.find(Icon).prop('src');

      // Expect
      expect(source).toEqual(VolumeOff);
    });

    it('Should show volume medium when the volume is between 1 and 49', () => {
      // Given
      const props = createDefaultProps({ mute: false, volume: 1 });
      const wrapper = shallow(<Volume {...props} />);

      // When
      const source = wrapper.find(Icon).prop('src');

      // Expect
      expect(source).toEqual(VolumeMedium);
    });

    it('Should show volume medium when the volume is between 50 and 100', () => {
      // Given
      const props = createDefaultProps({ mute: false, volume: 50 });
      const wrapper = shallow(<Volume {...props} />);

      // When
      const source = wrapper.find(Icon).prop('src');

      // Expect
      expect(source).toEqual(VolumeHigh);
    });
  });

  describe('Slider props', () => {
    it('Should pass 0 as value if we are mute', () => {
      // Given
      const props = createDefaultProps({ mute: true, volume: 1 });
      const wrapper = shallow(<Volume {...props} />);

      // When
      wrapper.find(BarWithPointer).simulate('click');
      const result = wrapper.find(Slider).prop('value');

      // Expect
      expect(result).toEqual(0);
    });

    it('Should pass volume as value', () => {
      // Given
      const props = createDefaultProps({ mute: false, volume: 42 });
      const wrapper = shallow(<Volume {...props} />);

      // When
      wrapper.find(BarWithPointer).simulate('click');
      const result = wrapper.find(Slider).prop('value');

      // Expect
      expect(result).toEqual(42);
    });
  });

  describe('Dropdown and Click', () => {
    it('Should disabled click on the bar when the mouse is over the Dropdown', () => {
      // Given
      const props = createDefaultProps({ mute: false, volume: 42 });
      const wrapper = shallow(<Volume {...props} />);

      // When
      wrapper.find(BarWithPointer).simulate('click');
      wrapper.find(DropdownContainer).simulate('mouseOver');
      wrapper.find(BarWithPointer).simulate('click');
      const result = wrapper.find(DropdownContainer).exists();

      // Expect
      expect(result).toEqual(true);
    });

    it('Should re-enable click on the bar when the mouse is out the Dropdown', () => {
      // Given
      const props = createDefaultProps({ mute: false, volume: 42 });
      const wrapper = shallow(<Volume {...props} />);

      // When
      wrapper.find(BarWithPointer).simulate('click');
      wrapper.find(DropdownContainer).simulate('mouseOver');
      wrapper.find(DropdownContainer).simulate('mouseOut');
      wrapper.find(BarWithPointer).simulate('click');
      const result = wrapper.find(DropdownContainer).exists();

      // Expect
      expect(result).toEqual(false);
    });
  });
});
