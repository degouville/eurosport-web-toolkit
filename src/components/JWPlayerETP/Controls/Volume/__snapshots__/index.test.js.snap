// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Volume Snapshots Should match snapshot when the volume is not shown 1`] = `
<Styled(div)
  innerRef={
    Object {
      "current": null,
    }
  }
  medium={true}
  onClick={[Function]}
>
  <Styled(img)
    alt="Volume"
    src="/eurosport-web-toolkit/da25f9b732454df2ec19e374f2219635.svg"
  />
</Styled(div)>
`;
